In our project, we have a "shared" folder inside "app" folder. And inside that, we have a "utils" folder which contains a "utils.ts" file. In this file, we have to functions exported.

Now, these are simply functions that are not related to Angular in general and we can use them in various places in the application.And it means, they are completely isolated of Angular.

And testing such utilities is the easiest way to test due to their isolation from Angular.

So, let's write some tests for this file. But where do we write them? In Angular, we follow a specific naming convention for test files. The files should have the same name as the file for which we are writing the tests, and it should end with "spec.ts".

It means, for "utils.ts", we will create a new file named "utils.spec.ts".

And we can now also run "jest" in "watch" mode because it means as we write tests in our utils.ts file, it will re run the tests automatically and show the progress in the console.

At the moment, if you run "npm run test:watch", it will show an error saying - 

    Your test suite must contain at least one test.

It is normal because our spec.ts file is empty and has no tests yet.

# WRITING OUR FIRST UNIT TEST

To start writing the tests, we first use the "describe" method that is provided by jest to break the test suite into different components. In simple words, we use it to group tests together.

For example, since we are testing the utils.ts file, we can write - 

    describe('utils', () => {
    
        // Tests here

    });

So now, we have this section named "utils" in which we will add all the tests that are related to "utils" file.

To write a test, we will use the "it" function. Just like "describe", the "it" function takes two arguments, where first is the name of the unit test and the second is a callback where we write the code for this unit test.

    describe('utils', () => {
    
        it('some name of the test', () => {
            expect(1).toEqual(1);
        });

    });

So here, we wrote a simple unit test which simply has this line of code - 

    expect(1).toEqual(1);

It means, in this unit test, we are expecting "1" to be equal to "1". Ofcourse this is true which means this test case will always pass. But, now that we have written this, if we save the file and check the console, we will see that the console now shows this - 

    utils
        √ some name of the test (11 ms)

It means, the unit test named "some name of the test" has passed.

# DESCRIBE INSIDE DESCRIBE

We can have describe functions inside describe functions. For example, we create a group for "utils" file using the outermost describe. Now, we also want to create a separate group for all the tests related to the "range()" function in utils.ts. And the same for the "pluck()" function.

It means, we will need two more describe() methods inside the outermost describe method.


    describe('utils', () => {
        
        describe('range', () => {
        });


        describe('pluck', () => {
        });

    });

And now, let's write some meaningful unit test.

# WRITING TESTS

Let's say we want to write a unit test for the "range" function so that if we call this function with "start" and "end" as 1 and 5, then it should return an array [1,2,3,4].

But, it is always recommended to alawys fail first. That means we need to prioritize failing test cases over nonfailing ones.

Why?

You write a failing test case in order to observe that the test case CAN fail.

It is to prove that the test itself is in fact doing its job.

If the test passes before you've written or changed any code then clearly the test is not very effective, so write the test, ensure that it fails, then write the code to satisfy the test. Really, with TDD, each piece of code you write should be to fix a failing test. This way you ensure that your code is fully tested.

So, let's write a test that will fail - 

    it('returns correct range from 1 to 5', () => {
        expect(range(1,5)).toEqual([1,1,1]);
    });

We know that the range function when passed "1" and "5" as start and end values should return [1,2,3,4]. But, in our test, we are expecting the output to be [1,1,1].

So, our test will fail. And we see in the console all the details of the failure.

It will show which test failed -

    utils › range › returns correct range from 1 to 5

It will also show what was the actual data returned by "range(1,5)". 

    - Expected  - 2
    + Received  + 3

      Array [
        1,
    -   1,
    -   1,
    +   2,
    +   3,
    +   4,
      ]

The above text means that we were expecting the output of the function to be [1,1,1] but it was [1,2,3,4]

So, now that we know what is the actual output, we can update our failing test to -

    it('returns correct range from 1 to 5', () => {
        expect(range(1,5)).toEqual([1,2,3,4]);
    });

And now, it will pass.

But, we should write as many tests as possible for a function so that we are testing all types of input values and see how the function behaves.

Anywyas, let's now write a unit test for the "pluck" function.

    describe('pluck', () => {
            it('returns corect result', () => {
                const data = [
                    {id: '1', name: 'foo'},
                    {id: '2', name: 'bar'},
                    {id: '3', name: 'baz'}
                ]
                expect(pluck(data, 'id')).toEqual([]);
            });
        });

Note that we have explicitly written the test so that it fails. Because we know that ideally, if we pass the above "data" array and we want to get the "id" field values, we should get an array as ["1", "2", "3"]. 

But, as mentioned above, we should fail-first. So, our test case fails, and it shows that the expected was ["1", "2", "3"]. So, we can update the test case accordingly.

    it('returns corect result', () => {
        const data = [
            {id: '1', name: 'foo'},
            {id: '2', name: 'bar'},
            {id: '3', name: 'baz'}
        ]
        expect(pluck(data, 'id')).toEqual(["1", "2", "3"]);
    });

And well, that's basically how we can test utilities in Angular. There is nothing fancy here because utilities are already isolated pieces of code so they are very easy to test.